src/anyio/_backends/_asyncio.py

```diff
@@ -270,13 +270,20 @@ def __exit__(
                 exc_val.exceptions if isinstance(exc_val, ExceptionGroup) else [exc_val]
             )
             if all(isinstance(exc, CancelledError) for exc in exceptions):
+
+                def uncancel() -> None:
+                    if hasattr(self._host_task, "uncancel"):
+                        self._host_task.uncancel()  # type: ignore[union-attr]
+
                 if self._timeout_expired:
+                    uncancel()
                     return True
                 elif not self._cancel_called:
                     # Task was cancelled natively
                     return None
```

Comment:
native cancellation on 3.11 can now be detected by comparing the number of times anyio called .cance() and the state of `task.cancelling()`

Reply:
I've added detection of native cancellation, and a test to that end, on OP's behalf.